### PR TITLE
feat: JobShell, Job Run, integrated Quartz.Net, implement FailedStrategy, TimeoutStrategy, BlockStrategy

### PR DESCRIPTION
**Pull request by gavin1ee**
implement JobShell, Job Run, integrated Quartz.Net, implement FailedStrategy, TimeoutStrategy, BlockStrategy

### CODE REVIEW

File: src/ApiGateways/Caller/Masa.Scheduler.ApiGateways.Caller/Services/SchedulerServerManagerService.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+using Masa.Scheduler.Contracts.Server.Model;
```
Comment (MayueCif) [inline src/ApiGateways/Caller/Masa.Scheduler.ApiGateways.Caller/Services/SchedulerServerManagerService.cs:4]: global using

File: src/Contracts/Masa.Scheduler.Contracts.Server/Infrastructure/IntegrationEvents/StartWaittingTaskIntergrationEvent.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Scheduler.Contracts.Server.Infrastructure.IntegrationEvents;
+
+public record StartWaittingTaskIntergrationEvent : IntegrationEvent
```
Comment (MayueCif) [inline src/Contracts/Masa.Scheduler.Contracts.Server/Infrastructure/IntegrationEvents/StartWaittingTaskIntergrationEvent.cs:6]: Waitting?

File: src/Web/Masa.Scheduler.Web.Admin/Pages/Teams/Components/SchedulerTasks.razor.cs [hunk 3]
```diff
@@ -263,5 +266,26 @@ private Task RadioGroupClickHandler()
...
+        if(Page > 1)
+        {
+            Page--;
+        }
+        return Task.CompletedTask;
+    }
+
+    private Task OnNextHandler()
+    {
+        Page++;
```
Comment (MayueCif) [inline src/Web/Masa.Scheduler.Web.Admin/Pages/Teams/Components/SchedulerTasks.razor.cs:281]: Upper limit？

File: src/Services/Masa.Scheduler.Services.Server/Application/Jobs/SchedulerJobCommandHandler.cs [hunk 4]
```diff
@@ -52,6 +68,8 @@ public async Task RemoveHandleAsync(RemoveSchedulerJobCommand command)
         }
 
         await _schedulerJobRepository.RemoveAsync(job);
+
+        await _quartzUtils.RemoveCronJob(job.Id);
```
Comment (MayueCif) [inline src/Services/Masa.Scheduler.Services.Server/Application/Jobs/SchedulerJobCommandHandler.cs:72]: move to QuartzUtils schedulerJobDomainService

File: src/Services/Masa.Scheduler.Services.Server/Domain/EventHandler/NotifyTaskRunResultDomainEventHandler.cs [hunk 5]
```diff
@@ -9,13 +9,19 @@ public class NotifyTaskRunResultDomainEventHandler
     private readonly SchedulerDbContext _dbContext;
     private readonly IRepository<SchedulerJob> _schedulerJobRepository;
     private readonly IHubContext<NotificationsHub> _hubContext;
+    private readonly IDistributedCacheClient _distributedCacheClient;
+    private readonly QuartzUtils _quartzUtils;
+    private readonly IIntegrationEventBus _eventBus;
 
-    public NotifyTaskRunResultDomainEventHandler(IRepository<SchedulerTask> schedulerTaskRepository, SchedulerDbContext dbContext, IRepository<SchedulerJob> schedulerJobRepository, IHubContext<NotificationsHub> hubContext)
+    public NotifyTaskRunResultDomainEventHandler(IRepository<SchedulerTask> schedulerTaskRepository, SchedulerDbContext dbContext, IRepository<SchedulerJob> schedulerJobRepository, IHubContext<NotificationsHub> hubContext, IDistributedCacheClient distributedCacheClient, QuartzUtils quartzUtils, IIntegrationEventBus eventBus)
```
Comment (MayueCif) [inline src/Services/Masa.Scheduler.Services.Server/Domain/EventHandler/NotifyTaskRunResultDomainEventHandler.cs:16]: to many params can format
image

File: src/Services/Masa.Scheduler.Services.Server/Domain/EventHandler/RegisterCronJobDomainEventHandler.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+    public RegisterCronJobDomainEventHandler(QuartzUtils quartzUtils)
+    {
+        _quartzUtils = quartzUtils;
+    }
+
+    [EventHandler]
+    public async Task RegisterCronJobAsync(RegisterCronJobDomainEvent @event)
+    {
+        if(@event.Request.Data.ScheduleType == ScheduleTypes.Cron && !string.IsNullOrEmpty(@event.Request.Data.CronExpression))
```
Comment (MayueCif) [inline src/Services/Masa.Scheduler.Services.Server/Domain/EventHandler/RegisterCronJobDomainEventHandler.cs:18]: IsNullOrWhiteSpace

File: src/Web/Masa.Scheduler.Web.Admin.Server/wwwroot/i18n/zh-CN.json [hunk 7]
```diff
@@ -46,6 +46,8 @@
   "Failure": "失败",
   "Timeout": "超时",
   "TimeoutSuccess": "超时成功",
```
Comment (MayueCif) [inline src/Web/Masa.Scheduler.Web.Admin.Server/wwwroot/i18n/zh-CN.json:48]: 超时(成功)

File: src/Web/Masa.Scheduler.Web.Admin/Pages/Teams/Components/SchedulerJobJobs.razor.cs [hunk 8]
```diff
@@ -436,4 +437,25 @@ private Task RadioGroupClickHandler()
...
+        if (Page > 1)
+        {
+            Page--;
+        }
+        return Task.CompletedTask;
+    }
+
+    private Task OnNextHandler()
+    {
+        Page++;
```
Comment (MayueCif) [inline src/Web/Masa.Scheduler.Web.Admin/Pages/Teams/Components/SchedulerJobJobs.razor.cs:452]: Upper limit？

File: src/Services/Masa.Scheduler.Services.Server/Domain/Managers/Servers/SchedulerServerManager.cs [hunk 9]
```diff
@@ -37,6 +45,46 @@ public override async Task OnManagerStartAsync()
         await base.OnManagerStartAsync();
 
         await StartAssignAsync();
+
+        await RegisterCronJob();
```
Comment (MayueCif) [inline src/Services/Masa.Scheduler.Services.Server/Domain/Managers/Servers/SchedulerServerManager.cs:49]: RegisterCronJobAsync

File: src/Services/Masa.Scheduler.Services.Server/Domain/Managers/Servers/SchedulerServerManager.cs [hunk 10]
```diff
@@ -37,6 +45,46 @@ public override async Task OnManagerStartAsync()
...
+
+        var allTask = await _dbContext.Tasks.Include(t => t.Job).Where(t => t.TaskStatus == TaskRunStatus.Running || t.TaskStatus == TaskRunStatus.WaitToRetry).ToListAsync();
+
+        await LoadRunningTask(allTask);
```
Comment (MayueCif) [inline src/Services/Masa.Scheduler.Services.Server/Domain/Managers/Servers/SchedulerServerManager.cs:53]: Async suffix

File: src/Services/Masa.Scheduler.Services.Worker/Domain/Managers/Workers/TaskHandlers/JobAppTaskHandler.cs [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+        catch (Exception ex)
+        {
+            _logger.LogError(ex, $"Process run Error, TaskId: {taskId}");
+            throw new UserFriendlyException($"Process run Error, TaskId: {taskId}");
+        }
+
+        return _runStatus;
+    }
+
+    private void JobApp_Exit(object? sender, EventArgs e)
```
Comment (MayueCif) [inline src/Services/Masa.Scheduler.Services.Worker/Domain/Managers/Workers/TaskHandlers/JobAppTaskHandler.cs:83]: remove _

File: src/Services/Masa.Scheduler.Services.Worker/Domain/Managers/Workers/TaskHandlers/JobAppTaskHandler.cs [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+            taskId.ToString(),
+            Path.Combine(resourcePath, dto.JobAppConfig.JobEntryAssembly),
+            dto.JobAppConfig.JobEntryMethod,
+            dto.JobAppConfig.JobParams
+        };
+
+        return string.Join(" ", parameterList);
+    }
+
+    private void JobApp_ErrorDataReceived(object? sender, System.Diagnostics.DataReceivedEventArgs e)
```
Comment (MayueCif) [inline src/Services/Masa.Scheduler.Services.Worker/Domain/Managers/Workers/TaskHandlers/JobAppTaskHandler.cs:102]: remove _

File: src/Services/Masa.Scheduler.Services.Worker/Domain/Managers/Workers/TaskHandlers/JobAppTaskHandler.cs [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+
+            var client = _httpClientFactory.CreateClient();
+
+            var response = await client.GetAsync(resource.FilePath);
+
+            response.EnsureSuccessStatusCode();
+
+            await using var stream = await response.Content.ReadAsStreamAsync();
+
+            await using var fs = File.Create(filePath);
```
Comment (MayueCif) [inline src/Services/Masa.Scheduler.Services.Worker/Domain/Managers/Workers/TaskHandlers/JobAppTaskHandler.cs:202]: await?

File: src/Shells/Masa.Scheduler.Shells.JobShell/Program.cs [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+{
+    assembly = Assembly.LoadFrom(path);
+}
+catch (Exception ex)
+{
+    result.Message = "Load Assembly Error, Exception Message: " + ex.Message;
+    return;
+}
+
+if(assembly == null)
```
Comment (MayueCif) [inline src/Shells/Masa.Scheduler.Shells.JobShell/Program.cs:34]: move to catch

**Review by MayueCif — COMMENTED** [review]
@gavin1ee

**Review by MayueCif — APPROVED** [review]
@doddgu

File: src/Services/Masa.Scheduler.Services.Server/Infrastructure/Quartz/QuartzUtils.cs [hunk 15]
```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Scheduler.Services.Server.Infrastructure.Quartz;
+
+public class QuartzUtils
+{
+    private readonly ISchedulerFactory _schedulerFactory;
+    private IScheduler _scheduler = default!;
+    public QuartzUtils(ISchedulerFactory schedulerFactory)
```
Comment (doddgu) [inline src/Services/Masa.Scheduler.Services.Server/Infrastructure/Quartz/QuartzUtils.cs:10]: add blank line.

File: src/Services/Masa.Scheduler.Services.Server/Domain/QuartzJob/StartSchedulerTaskQuartzJob.cs [hunk 16]
```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Scheduler.Services.Server.Domain.QuartzJob;
+
+public class StartSchedulerTaskQuartzJob : IJob
+{
+    private readonly IDomainEventBus _eventBus;
+    public StartSchedulerTaskQuartzJob(IDomainEventBus eventBus)
```
Comment (doddgu) [inline src/Services/Masa.Scheduler.Services.Server/Domain/QuartzJob/StartSchedulerTaskQuartzJob.cs:9]: add blank line.

File: src/Services/Masa.Scheduler.Services.Server/Domain/QuartzJob/StartSchedulerJobQuartzJob.cs [hunk 17]
```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) MASA Stack All rights reserved.
+// Licensed under the Apache License. See LICENSE.txt in the project root for license information.
+
+namespace Masa.Scheduler.Services.Server.Domain.QuartzJob;
+
+public class StartSchedulerJobQuartzJob : IJob
+{
+    private readonly IDomainEventBus _eventBus;
+    public StartSchedulerJobQuartzJob(IDomainEventBus eventBus)
```
Comment (doddgu) [inline src/Services/Masa.Scheduler.Services.Server/Domain/QuartzJob/StartSchedulerJobQuartzJob.cs:9]: add blank line.

File: src/Services/Masa.Scheduler.Services.Server/Properties/launchSettings.json [hunk 18]
```diff
@@ -3,8 +3,8 @@
     "windowsAuthentication": false,
     "anonymousAuthentication": true,
     "iisExpress": {
-      "applicationUrl": "http://localhost:46239",
-      "sslPort": 46539
+      "applicationUrl": "http://localhost:50672/",
```
Comment (doddgu) [inline src/Services/Masa.Scheduler.Services.Server/Properties/launchSettings.json:6]: See the port rule from confluence.

**Review by doddgu — CHANGES_REQUESTED** [review]

**Review by doddgu — APPROVED** [review]
LGTM

**Pull request merged**